many: switch to apparmor 5.x with 5 ABI

zyga opened the pull request:
This is extremely early work as we need to align some stars before we get to hit the bugs and fix them:

1) ~~we need apparmor 5.0.0 alpha 2 for upstream af_unix features~~
2) ~~we need to actively use abi 5 across our profiles~~
3) interfaces will likely need some changes
4) ~~there's no abi 5.0 file in the release we are using so that part of the code is dormant~~

Then we get to test and see what breaks.

For apparmor master with 5 abi please see: https://github.com/canonical/snapd/pull/16780
For apparmor 5.x with 5 ABI please see: https://github.com/canonical/snapd/pull/15967
For apparmor 5.x with 4 ABI please see: https://github.com/canonical/snapd/pull/16781

zyga committed:
sandbox/apparmor: use abi 5.0 if possible

Use 5.0 ABI when available. This may affect a few profiles so it's
likely to be coupled with additional changes after the first round of
testing.

Jira: https://warthogs.atlassian.net/browse/SNAPDENG-35413

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

zyga committed:
sandbox/apparmor: refactor internal parser test setup

Extract duplicated internal apparmor parser fixture setup and assertions into
shared test helpers. Rework TestAppArmorInternalAppArmorParserAbi3/4 to use
helpers and add ABI 5 coverage with TestAppArmorInternalAppArmorParserAbi5.
Reuse the helper in TestInternalParser and
TestSetupConfCacheDirsWithInternalApparmor to remove duplicate fixture code.
Rename helper to setupInternalAppArmorParserEnv for clearer scope.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

zyga committed:
tests: expect apparmor 5 in snap-debug-execution-apparmor

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

bboozzoo commented on tests/main/debug-execution/task.yaml:
```diff
@@ -115,7 +115,7 @@ execute: |
                 MATCH 'apparmor-parser: /snap/snapd/.*/usr/lib/snapd/apparmor_parser' < snap-apparmor-force-reexec.out
```
would it make sense to assert `apparmor-parser-command` here as well? the trouble is that it would likely need another `case "$SPREAD_SYSTEM"` to account for different versions of ABIs on the host across releases?

bboozzoo replied:
actually this would also be a record of what ABI is used across Ubuntu releases if we happen to run from the native package.

zyga replied:
I'll look into that

zyga replied:
Done

bboozzoo commented on tests/main/debug-execution/task.yaml:
```diff
@@ -110,21 +110,57 @@ execute: |
+                        MATCH 'apparmor-parser-command: (/usr)?/sbin/apparmor_parser$' < snap-apparmor-default.out
+                        ;;
+                    ubuntu-22.04-64|ubuntu-24.04-64|ubuntu-26.04-64|debian-11-64|debian-12-64)
+                        MATCH 'apparmor-parser-command: (/usr)?/sbin/apparmor_parser --policy-features (/usr)?/etc/apparmor\.d/abi/3\.0' < snap-apparmor-default.out
```
we're still using abi 4 on 26.04?

ZeyadYasser replied:
```shell
$ SNAP_REEXEC=0 snap debug execution apparmor
apparmor-parser: /usr/sbin/apparmor_parser
apparmor-parser-command: /usr/sbin/apparmor_parser --policy-features /etc/apparmor.d/abi/3.0
internal: false
$ ls /etc/apparmor.d/abi/
3.0  4.0  4.0-ip  5.0  kernel-5.4-outoftree-network  kernel-5.4-vanilla
```

even 5.0 exists but they intentionally ignored by:
https://github.com/canonical/snapd/blob/8c923238199c5694e4e8de1a76ffba67bc2083a4/sandbox/apparmor/apparmor.go#L947-L956

ZeyadYasser commented on sandbox/apparmor/apparmor.go:
I wonder if we should also do the same for 5.0 for completeness

zyga replied:
yes, good idea

zyga replied:
Done

ZeyadYasser commented on sandbox/apparmor/apparmor.go:
for completeness
```suggestion
hostAbi30File = filepath.Join(newrootdir, "/etc/apparmor.d/abi/3.0")
hostAbi40File = filepath.Join(newrootdir, "/etc/apparmor.d/abi/4.0")
hostAbi50File = filepath.Join(newrootdir, "/etc/apparmor.d/abi/5.0")
```

zyga replied:
Done

ZeyadYasser reviewed (approved):
LGTM, Thank you! left a comment on extra debug logs for completeness.

zyga committed:
sandbox/apparmor: add test coverage for distro parser ABI detection

Add a table-driven test covering AppArmorParser() behavior in the distro
parser search path when host ABI files are present. Previously only the
internal apparmor_parser path was tested with mock ABI files; the distro
path relied on unmockable global variables.

Three new export_test.go mocks (MockHostAbi30File, MockHostAbi40File,
MockHostAbi50File) enable testing of the detection-and-ignore logic for
ABI 4.0 and ABI 5.0 that coexists with pinning to ABI 3.0 when available.

Five subtest cases verify: no ABI files (plain parser), only ABI 3.0
(pinned), only ABI 5.0 ignored, both ABI 4+5 ignored, and all three ABIs
present (pins to 3.0, ignores 4/5). The new test also asserts on the debug
log output via logger.MockLogger() for cases that exercise the detection.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

bboozzoo commented on sandbox/apparmor/apparmor.go:
```diff
@@ -944,6 +951,14 @@ func AppArmorParser() (cmd *exec.Cmd, internal bool, err error) {
 		if _, err := os.Stat(path); err == nil {
 			logger.Debugf("checking distro apparmor_parser at %v", path)
+
+			// Detect bug ignore apparmor 5.0 ABI support.
```
```suggestion
			// Detect but ignore apparmor 5.0 ABI support.
```

zyga replied:
Thanks!

bboozzoo commented on sandbox/apparmor/apparmor.go:
```diff
@@ -944,6 +951,14 @@ func AppArmorParser() (cmd *exec.Cmd, internal bool, err error) {
+			// Detect bug ignore apparmor 5.0 ABI support.
+			// There's no known, diagnosed issue with ABI 5 yet but given that
+			// we also ignore host ABI 4.0, it's sensible to ignore this one explicitly too.
+			if fi, err := os.Lstat(hostAbi50File); err == nil && !fi.IsDir() {
+				logger.Debugf("apparmor 5.0 ABI detected but ignored")
```
shouldn't this be the same as if we used abi5.0 from the snapd snap?

zyga replied:
That's a valid question. Right now I would keep host apparmor to "let's not poke the bear" mode until we decide when and how we use host apparmor features.

I'll prepare a 2nd PR that lifts both restrictions so that we can see how we do but I worry that our test coverage in this area is not great. I need to see how to trigger a run that mostly uses the deb.

bboozzoo reviewed (approved):
LGTM

zyga committed:
sandbox/apparmor: fix typo

Co-authored-by: Maciej Borzecki <maciek.borzecki@gmail.com>
